index.d.ts

```diff
@@ -29,158 +32,10 @@ import got = require('got');
-			pAll.PromiseFactory<Result6>,
-			pAll.PromiseFactory<Result7>,
-			pAll.PromiseFactory<Result8>,
-			pAll.PromiseFactory<Result9>
-		],
-		options?: pAll.Options
-	): Promise<
-	[
-		Result1,
-		Result2,
-		Result3,
-		Result4,
-		Result5,
-		Result6,
-		Result7,
-		Result8,
-		Result9
-	]
-	>;
-	<Result1, Result2, Result3, Result4, Result5, Result6, Result7, Result8>(
-		tasks: [
-			pAll.PromiseFactory<Result1>,
-			pAll.PromiseFactory<Result2>,
-			pAll.PromiseFactory<Result3>,
-			pAll.PromiseFactory<Result4>,
-			pAll.PromiseFactory<Result5>,
-			pAll.PromiseFactory<Result6>,
-			pAll.PromiseFactory<Result7>,
-			pAll.PromiseFactory<Result8>
-		],
-		options?: pAll.Options
-	): Promise<
-	[Result1, Result2, Result3, Result4, Result5, Result6, Result7, Result8]
-	>;
-	<Result1, Result2, Result3, Result4, Result5, Result6, Result7>(
-		tasks: [
-			pAll.PromiseFactory<Result1>,
-			pAll.PromiseFactory<Result2>,
-			pAll.PromiseFactory<Result3>,
-			pAll.PromiseFactory<Result4>,
-			pAll.PromiseFactory<Result5>,
-			pAll.PromiseFactory<Result6>,
-			pAll.PromiseFactory<Result7>
-		],
-		options?: pAll.Options
-	): Promise<[Result1, Result2, Result3, Result4, Result5, Result6, Result7]>;
-	<Result1, Result2, Result3, Result4, Result5, Result6>(
-		tasks: [
-			pAll.PromiseFactory<Result1>,
-			pAll.PromiseFactory<Result2>,
-			pAll.PromiseFactory<Result3>,
-			pAll.PromiseFactory<Result4>,
-			pAll.PromiseFactory<Result5>,
-			pAll.PromiseFactory<Result6>
-		],
-		options?: pAll.Options
-	): Promise<[Result1, Result2, Result3, Result4, Result5, Result6]>;
-	<Result1, Result2, Result3, Result4, Result5>(
-		tasks: [
-			pAll.PromiseFactory<Result1>,
-			pAll.PromiseFactory<Result2>,
-			pAll.PromiseFactory<Result3>,
-			pAll.PromiseFactory<Result4>,
-			pAll.PromiseFactory<Result5>
-		],
-		options?: pAll.Options
-	): Promise<[Result1, Result2, Result3, Result4, Result5]>;
-	<Result1, Result2, Result3, Result4>(
-		tasks: [
-			pAll.PromiseFactory<Result1>,
-			pAll.PromiseFactory<Result2>,
-			pAll.PromiseFactory<Result3>,
-			pAll.PromiseFactory<Result4>
-		],
-		options?: pAll.Options
-	): Promise<[Result1, Result2, Result3, Result4]>;
-	<Result1, Result2, Result3>(
-		tasks: [
-			pAll.PromiseFactory<Result1>,
-			pAll.PromiseFactory<Result2>,
-			pAll.PromiseFactory<Result3>
-		],
-		options?: pAll.Options
-	): Promise<[Result1, Result2, Result3]>;
-	<Result1, Result2>(
-		tasks: [pAll.PromiseFactory<Result1>, pAll.PromiseFactory<Result2>],
-		options?: pAll.Options
-	): Promise<[Result1, Result2]>;
-	<Result1>(
-		tasks: [pAll.PromiseFactory<Result1>],
-		options?: pAll.Options
-	): Promise<[Result1]>;
-	<TAll>(
-		tasks: Iterable<pAll.PromiseFactory<TAll>> | Array<pAll.PromiseFactory<TAll>>,
-		options?: pAll.Options
-	): Promise<TAll[]>;
+	<Task extends pAll.PromiseFactory<unknown>[]>(
+		tasks: readonly [...Task],
+		options?: pAll.Options,
+	): Promise<{ [P in keyof Task]: Awaited<ReturnType<Task[P]>> }>
```

Comment:
Getting a TS error on the `Task[P]`:

```
Type 'Task[P]' does not satisfy the constraint '(...args: any) => any'.
  Type 'Task[keyof Task]' is not assignable to type '(...args: any) => any'.
    Type 'Task[string] | Task[number] | Task[symbol]' is not assignable to type '(...args: any) => any'.
      Type 'Task[string]' is not assignable to type '(...args: any) => any'.ts(2344)
```

I'm not sure how to resolve this. Would love any help on this.

Reply:
`pAll` returns a promise for an array of fulfilled values. `{ [P in keyof Task]: Awaited<ReturnType<Task[P]>> }>` is not an array.

Reply:
The type tests are passing so it should be returning an array of fulfilled values. Not to mention, I think it's necessary to use the `{ [P in key of Array]: Array[P] }` syntax to map over the array.


I also asked a similar question (related but different problem) on Stackoverflow. Hopefully will find a way to resolve this soon.

Reply:
Got it working! Just needed to make sure that `Task[P]` was a function. Simple solution 🤦‍♀️